backend/src/memberships/memberships.controller.ts

```diff
@@ -9,27 +9,121 @@ import {
+import { EntityNotFoundError } from 'typeorm';
+import * as bcrypt from 'bcrypt';
+import { Request } from 'express';
+import { User } from 'src/users/entities/user.entity';
+
+async function userIsAdmin(
+  userId: number,
+  channelId: number,
+  membershipsService: MembershipsService,
+): Promise<boolean> {
+  const creatorMembership = await membershipsService.findAll({
+    user: userId.toString(),
+    channel: channelId.toString(),
+  });
+  if (
+    creatorMembership.length === 1 &&
+    (creatorMembership[0].role === MembershipRoleType.ADMIN ||
+      creatorMembership[0].role === MembershipRoleType.OWNER)
+  ) {
+    return true;
+  }
+  return false;
+}
+
+async function userIsOwner(
+  userId: number,
+  channelId: number,
+  membershipsService: MembershipsService,
+): Promise<boolean> {
+  const creatorMembership = await membershipsService.findAll({
+    user: userId.toString(),
+    channel: channelId.toString(),
+  });
+  if (
+    creatorMembership.length === 1 &&
+    creatorMembership[0].role === MembershipRoleType.OWNER
+  ) {
+    return true;
+  }
+  return false;
+}
 
 @ApiBearerAuth()
 @ApiTags('channel memberships')
 @Controller('memberships')
 export class MembershipsController {
-  constructor(private readonly membershipsService: MembershipsService) {}
+  constructor(
+    private readonly membershipsService: MembershipsService,
+    private readonly channelsService: ChannelsService,
+  ) {}
 
   @Post()
   async create(
     @Body() createMembershipDto: CreateMembershipDto,
+    @Req() req: Request,
   ): Promise<ResponseMembershipDto> {
+    try {
+      let isAuthorized = false;
+      const channel = await this.channelsService.findOne(
+        createMembershipDto.channelId,
+      );
+      if (createMembershipDto.role === MembershipRoleType.OWNER) {
+        throw new UnauthorizedException();
+      }
+      if (createMembershipDto.role === MembershipRoleType.ADMIN) {
+        if (
+          !(await userIsOwner(
+            (req.user as User).id,
+            channel.id,
+            this.membershipsService,
+          ))
+        ) {
+          throw new UnauthorizedException();
+        }
+      }
+      if (channel.type === ChannelType.PRIVATE) {
+        isAuthorized = await userIsAdmin(
+          (req.user as User).id,
+          createMembershipDto.channelId,
+          this.membershipsService,
+        );
+      } else if (channel.type === ChannelType.PROTECTED) {
+        isAuthorized =
+          createMembershipDto.password &&
+          (await bcrypt.compare(
+            createMembershipDto.password,
+            channel.password,
+          ));
+      } else {
+        isAuthorized = true;
+      }
+      if (!isAuthorized) {
+        throw new UnauthorizedException();
+      }
+    } catch (error) {
+      if (error instanceof EntityNotFoundError) {
+        throw new BadRequestException('Cannot find channel');
+      }
+      throw error;
+    }
```

Comment:
There's a lot of checks here which relate to other services, wouldn't it be better to put most of them in membership service ? Are they here because of the request interaction ?

Reply:
I agree that it is a lot of checks, but yes, the thinking was that logic related to the request (and therefore the logged-in user) should be handled by the controller. This way the service is focused solely on interacting with the DB and it is the controller which determines if the service can be used. One advantage of this being that we can use the service elsewhere without having to worry about creating a request/user who is allowed to use it. (For example, when creating a new channel it would be a good idea for the backend to create the owner membership automatically, or even just for running unit tests)